[NFC][analyzer] Document configuration options

NagyDonat opened the pull request:
This commit documents the process of specifying values for the analyzer options and checker options implemented in the static analyzer, and adds a script which includes the documentation of the analyzer options (which was previously only available through a command-line flag) in the RST-based web documentation.

NagyDonat committed:
[NFC][analyzer] Document configuration options

This commit documents the process of specifying values for the analyzer
options and checker options implemented in the static analyzer, and adds
a script which includes the documentation of the analyzer options (which
was previously only available through a command-line flag) in the
RST-based web documentation.

NagyDonat commented:
The HTML page produced by this commit is available at: https://gist.github.com/NagyDonat/8d8ca3ac2292ceb8044a4bd1b54f3a14

(I slightly edited the locally built HTML to ensure that it loads CSS/JS from clang.llvm.org/docs instead of local files.)

steakhal commented on clang/docs/CMakeLists.txt:
```diff
@@ -143,6 +143,32 @@ if (LLVM_ENABLE_SPHINX)
+    set(generate_aopts_docs generate_analyzer_options_docs.py)
+    add_custom_command(
+      OUTPUT ${aopts_rst}
+      COMMAND ${Python3_EXECUTABLE} ${generate_aopts_docs} -i ${analyzeroptions_def} -t ${aopts_rst_in} -o ${aopts_rst}
```
BTW isn't the `-i` flag meant for interactive uses?

NagyDonat replied:
This is unrelated to the interpreter flag: the Python interpreter is parametrized as
```
python3 <arguments for the interpreter> script.py <arguments for the script>
```
and I decided to implement an `-i` flag for specifying the input file of the script. I could change this (probably to `--input`) if you think that it's misleading.

NagyDonat replied:
The names of the arguments of the script were clarified in https://github.com/llvm/llvm-project/pull/135169/commits/4fae8627073111934583f885ba23d7edf478bbbc

steakhal commented on clang/docs/analyzer/user-docs/Options.rst.in:
```diff
@@ -0,0 +1,102 @@
+With the clang driver
+---------------------
+
+In a conventional workflow ``clang -cc1`` (which is a low-level internal
+interface) is invoked indirectly by the clang *driver* (i.e. plain ``clang``
+without the ``-cc1`` flag), which acts as an "even more frontend" wrapper layer
+around the ``clang -cc1`` *compiler frontend*. In this situation **each**
+command line argument intended for the *compiler frontend* must be prefixed
+with ``-Xclang``.
```
We should put a disclaimer here that clang does not intend to preserve backward compatibility or announce breaking changes with `-cc1` flags.
Use them with that in mind.

steakhal replied:
BTW I believe that one could pass analyzer options to the analyzer action using the `-Xanalyzer` although I rarely ever see that.

NagyDonat replied:
OK, I'll add a disclaimer and perhaps document `-Xanalyzer` if that works as well.

steakhal commented on clang/docs/analyzer/user-docs/Options.rst.in:
```diff
@@ -0,0 +1,102 @@
+List of analyzer options
+========================
+
+.. warning::
+   These options are primarily intended for development purposes. Changing
+   their values may drastically alter the behavior of the analyzer, and may
+   even result in instabilities or crashes!
```
Generally, anything that is not the default is practically unsupported unfortunately. It's a messed up situation, but let's admit that. There are only a few exceptional configs, such as Z3 cross checking, but generally, this is the case.

steakhal replied:
Maybe we should say that crash reports are welcomed, and depending on the severity we may decide to fix them.

NagyDonat replied:
I extended this warning according to your suggestions in https://github.com/llvm/llvm-project/pull/135169/commits/461d3db784bd66574c9d175dbc8f5e7190342b8e

steakhal commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -0,0 +1,242 @@
+    )
+
+
+def tokenize(s):
+    result = []
+    pos = 0
+    while pos < len(s):
+        for regex, kind in TOKENS:
+            if m := regex.match(s, pos):
+                if kind is not None:
+                    result.append(Token(kind, m.group(0)))
+                pos = m.end()
+                break
+        else:
+            report_unexpected(s, pos)
+            pos += 1
+    return result
```
Do you intentionally tokenize yourself? I figured Python had dozens of libraries doing this for us.
Did you try to avoid 3rd party deps by this?

NagyDonat replied:
I didn't try to find an external tokenizer library, because `AnalyzerOptions.def` uses some tricky features (apostrophes in numeric literals, preprocessor directives continued with backslash at EOL), so I would need to use a reasonably feature-complete heavyweight C++-specific tokenizer, and interfacing with that wouldn't be much simpler than the current implementation.

Also I like that this little self-contained tokenizer is very transparent (if it misbehaves, the developer only needs to understand a few dozen lines) and I don't expect significant changes in the structure of `AnalyzerOptions.def` so I don't think that we need to prepare with a general-purpose tokenizer for future changes.

steakhal commented on clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:
```diff
@@ -7,6 +7,9 @@
 //===----------------------------------------------------------------------===//
 //
 //  This file defines the analyzer options avaible with -analyzer-config.
+//  Note that clang/docs/tools/generate_analyzer_options_docs.py relies on the
+//  structure of this file, so if this file is refactored, then make sure to
+//  update that script as well.
```
I wonder if we can have a test for this? At least some canary test that the script works.

NagyDonat replied:
In https://github.com/llvm/llvm-project/pull/135169/commits/180afc02b5f1ae640c892fa7df22984daff9c2d0 I added a test which validates that the script does understand `AnalyzerOptions.def`.

steakhal commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -0,0 +1,242 @@
+
+
+p = argparse.ArgumentParser()
+p.add_argument("-i", "--input", help="path to AnalyzerOptions.def")
+p.add_argument("-t", "--template", help="path of template file")
+p.add_argument("-o", "--output", help="path of output file")
```
AH!, so the `-i` wasn't the flag of `python` itself. You could use the verbose names of these flags in CMake then to remove this confusion.

steakhal commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -0,0 +1,242 @@
+    return result
+
+
+p = argparse.ArgumentParser()
```
I don't write "official" python scripts, but I've seen a pattern having a check for "__main__" and what not elsewhere - except in this script. Do you know what conventions to follow? Have you checked similar in-tree python scripts?

NagyDonat replied:
I copied this structure from another (somewhat analogous) script in the repo, but I didn't count the number of scripts that check / don't check `__main__`.

NagyDonat replied:
The LLVM repository contains 213 executable python scripts, and among them 135 (= 63%) check for `if __name__ == "__main__":`.

Based on this I'd say that both using and not using that pattern are acceptable, and based on personal preference I'm not using it here. I feel this pattern is only relevant if the same Python file is used both as a standalone script (when this `if` block executes) and a module that can be imported by other source files (when the `if` is skipped) -- and this file is only used as a standalone script.

steakhal commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -0,0 +1,242 @@
+            + desc_to_rst_paragraphs(desc)
+            + defaults_to_rst_paragraph(defaults)
+        )
+    except ValueError as ve:
+        print(ve.args[0], file=sys.stderr)
+        return ""
```
So we have some error handling. I wonder if we should prepare for (thus detect) some malformed `.def` files for common mistakes to see how that use-case would look like.
I think devs should have a clear message what did they mess up in a `.def` file.
We could demonstrate the proper detection in a test file.

NagyDonat replied:
Yep, I implemented some "fail gracefully" printouts for the situations where the script runs into an unexpected situation, but I wouldn't say that these are supported features of the script. As this code is only intended for one very specific non-user-facing role, I think even removing the error handling is more reasonable than writing tests for it.

NagyDonat replied:
I still feel that testing the error reporting logic of the script (with an input where the script is expected to fail) would be an overkill, because it is simple, straightforward, and I don't expect that further changes would break it accidentally.

On the other hand, I added some checks which would cause a test failure if changes in `AnalyzerOptions.def` would make obsolete one of the content tweaks in the script -- because that part of the script *is* very fragile.

If you think that there are other parts of the code that could easily and silently break during normal evolution of the code, tell me and I'll add tests for them.

steakhal reviewed (commented):
Thank you for proposing this. It looks really good already.
The code looks nice and dense, and the output also looks wonderful.
I had some design questions and a few remarks on the testing and error handling.

Good job.

NagyDonat commented on clang/docs/CMakeLists.txt:
```diff
@@ -143,6 +143,32 @@ if (LLVM_ENABLE_SPHINX)
+              copy-clang-rst-docs
+      )
+    add_custom_target(generate-analyzer-options-rst DEPENDS ${aopts_rst})
+    foreach(target ${docs_targets})
+      add_dependencies(${target} generate-analyzer-options-rst)
+    endforeach()
```
In this foreach loop _all the doc targets_ means two doc targets (or less if they aren't both enabled), and they both format the same set of RST files (the whole clang documentation) into different output formats. As the RST file generated by my script fits into that collection RST of files, it should be a prerequisite of both of those targets.

NagyDonat commented on clang/docs/CMakeLists.txt:
```diff
@@ -143,6 +143,32 @@ if (LLVM_ENABLE_SPHINX)
+    set(aopts_rst "${CMAKE_CURRENT_BINARY_DIR}/${aopts_rst_rel_path}")
+    set(analyzeroptions_def "${CMAKE_CURRENT_SOURCE_DIR}/../include/clang/StaticAnalyzer/Core/AnalyzerOptions.def")
+    set(aopts_rst_in "${CMAKE_CURRENT_SOURCE_DIR}/${aopts_rst_rel_path}.in")
+    set(generate_aopts_docs generate_analyzer_options_docs.py)
```
Fair point, I'll inline the script name.

NagyDonat replied:
Done in https://github.com/llvm/llvm-project/pull/135169/commits/c9ade009da319d7e7bec336eaa6d507bff5a3bfd

NagyDonat committed:
Inline the name of the script

NagyDonat committed:
Clarify argument names of the script generating the docs

NagyDonat committed:
Extend some disclaimers

NagyDonat commented:
@steakhal I implemented most of your review suggestions (the only missing thing is that I'll set up a test to validate that this script can parse `AnalyzerOptions.def`).

Please check the wording of the disclaimers added by https://github.com/llvm/llvm-project/pull/135169/commits/461d3db784bd66574c9d175dbc8f5e7190342b8e because I'm not too confident about them. (In particular, you suggested that I should add a disclaimer about the malleable state of `clang -cc1` flags, but there is no more stable alternative for specifying analyzer flags _and_ `-analyzer-config` is unchanged since 2017, so I also added an "anti-disclaimer" sentence. There is probably a better solution than this...)

NagyDonat committed:
Test generate_analyzer_options_docs.py

NagyDonat commented:
@steakhal Gentle ping.

steakhal commented on clang/docs/analyzer/user-docs/Options.rst.in:
```diff
@@ -0,0 +1,115 @@
+
+Here ``[...]`` stands for dozens of low-level flags which ensure that ``clang
+-cc1`` does the right thing (e.g. ``-fcolor-diagnostics`` when it's suitable;
+``-analyzer-checker`` flags to enable a sane default set of checkers). Also
```
```suggestion
``-analyzer-checker`` flags to enable the default set of checkers). Also
```

steakhal commented on clang/docs/analyzer/user-docs/Options.rst.in:
```diff
@@ -0,0 +1,115 @@
+Here ``[...]`` stands for dozens of low-level flags which ensure that ``clang
+-cc1`` does the right thing (e.g. ``-fcolor-diagnostics`` when it's suitable;
+``-analyzer-checker`` flags to enable a sane default set of checkers). Also
+note the distinction that the ``clang`` *driver* requires ``--analyze`` (double
+dashes) while the ``clang -cc1`` *compiler frontend* requires ``-analyze``
+(single dash).
```
I'm not sure if it was mentioned explicitly, but one should prefer `clang --analyze ...` over `-cc1 ...`.

NagyDonat replied:
Good point, I'll check.

steakhal commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -0,0 +1,286 @@
+def desc_to_rst_paragraphs(tok):
+    base_desc = string_value(tok)
+
+    # Escape a star that would act as inline emphasis within RST.
+    desc = base_desc.replace("ctu-max-nodes-*", r"ctu-max-nodes-\*")
+    if desc != base_desc:
+        err_handler.record_use_of_tweak("ctu-max-nodes-*")
```
If this is about escaping, then couldn't we just replace `*` characters with `\*` without knowing the string itself?

NagyDonat replied:
Yes, this is about escaping, and I'll follow your suggestion.

steakhal replied:
I think it's worth it. We don't need to go far and complete. If it's good enough then it's good for us.

NagyDonat replied:
I added code that always escapes `*` and `_`, and I think this is sufficient for avoiding accidental RST markup. (I also tried adding code that escapes backslashes, but there are no backslashes in the current documentation, so I decided against it -- however, I can add it if you strongly support it.)

steakhal replied:
Thank you!

steakhal commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -0,0 +1,286 @@
+    if desc != base_desc:
+        err_handler.record_use_of_tweak("ctu-max-nodes-*")
+
+    # Many descriptions end with "Value: <list of accepted values>", which is
+    # OK for a terse command line printout, but should be prettified for web
+    # documentation.
+    # Moreover, the option ctu-invocation-list shows some example file content
+    # which is formatted as a preformatted block.
+    paragraphs = [desc]
+    extra = ""
+    if m := re.search(r"(^|\s)Value:", desc):
+        err_handler.record_use_of_tweak("accepted values")
+        paragraphs = [desc[: m.start()], "Accepted values:" + desc[m.end() :]]
+    elif m := re.search(r"\s*Example file.content:", desc):
+        err_handler.record_use_of_tweak("example file content")
+        paragraphs = [desc[: m.start()]]
+        extra = "Example file content::\n\n  " + desc[m.end() :] + "\n\n"
```
Maybe we could mandate this and hard-error if this format is not met.
Have you thought about that?

NagyDonat replied:
`Value:` lists are only relevant and present in a minority of the descriptions, while currently there is just one option that specifies `Example file content`, so we cannot mandate that these should be present in _each_ description. The `record_use_of_tweak` calls only validate that _at least one description_ uses that particular tweak.

NagyDonat replied:
> However, for the "Value:" I don't see a drawback for mandating it.

I'll consider implementing it in a follow-up commit, but I don't want to touch the content of `AnalyzerOptions.def` in this commit.

steakhal commented:
Sorry for my delay. I missed the review request.

NagyDonat committed:
Apply simple suggestions from code review

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

NagyDonat commented on clang/docs/tools/generate_analyzer_options_docs.py:
```diff
@@ -227,8 +263,11 @@ def get_option_list(input_file):
-p.add_argument("--out", help="path of output file")
+p.add_argument("--template", help="template file")
+p.add_argument("--out", help="output file")
+p.add_argument(
+    "--validate", action="store_true", help="exit with failure on parsing error"
+)
```
I didn't enable validation when the script is run from `cmake` to compile the documentation because the script is robust enough to produce useful output even if it cannot recognize a few characters within `AnalyzerOptions.def`. However – now that I think about it – this is actually an unlikely situation, because changes to `AnalyzerOptions.def` would be caught by the tests.

To simplify the script I'll remove this `--validate` flag and always exit with failure when there is a parsing error.

steakhal reviewed (changes requested):
I'm just leaving a checkpoint for myself that I've reviewed this PR so far.

NagyDonat committed:
Tweak disclaimer

NagyDonat committed:
Generalize escaping RST special characters

NagyDonat committed:
Make the --validate flag always active

NagyDonat committed:
Fix code formatter issue

NagyDonat commented:
If I'm not mistaken I handled every review comment. Thanks for all the useful suggestions!

NagyDonat committed:
Report unexpected argument count as error

steakhal reviewed (approved):
:rocket:

NagyDonat commented:
This change only affects the documentation, the C++ assertion failure detected by the buildbot is completely unrelated.
